rule/struct_tag.go

```diff
@@ -158,286 +235,292 @@ func (lintStructTagRule) getTagName(tag *structtag.Tag) string {
 	}
 }
 
-// checkTaggedField checks the tag of the given field.
-// precondition: the field has a tag
-func (w lintStructTagRule) checkTaggedField(f *ast.Field) {
-	if len(f.Names) > 0 && !f.Names[0].IsExported() {
-		w.addFailure(f, "tag on not-exported field "+f.Names[0].Name)
+func checkASN1Tag(checkCtx *checkContext, tag *structtag.Tag, fieldType ast.Expr) (message string, succeeded bool) {
+	checkList := append(tag.Options, tag.Name)
+	for _, opt := range checkList {
+		switch opt {
+		case "application", "explicit", "generalized", "ia5", "omitempty", "optional", "set", "utf8":
+			// do nothing
+		default:
+			msg, ok := checkCompoundANS1Option(checkCtx, opt, fieldType)
+			if !ok {
+				return msg, false
+			}
+		}
 	}
 
-	tags, err := structtag.Parse(strings.Trim(f.Tag.Value, "`"))
-	if err != nil || tags == nil {
-		w.addFailure(f.Tag, "malformed tag")
-		return
-	}
+	return "", true
+}
 
-	for _, tag := range tags.Tags() {
-		if msg, ok := w.checkTagNameIfNeed(tag); !ok {
-			w.addFailure(f.Tag, msg)
+func checkCompoundANS1Option(checkCtx *checkContext, opt string, fieldType ast.Expr) (message string, succeeded bool) {
+	parts := strings.Split(opt, ":")
+	switch parts[0] {
```

Comment:
Here also, using strings.Cut would help for readability

It would avoid code like


                tagNumber := strings.TrimLeft(opt, "tag:")


                if len(parts) < 2 { 
                         return "malformed default", false 
                 }


And the use of parts[0] and parts[1]